### PR TITLE
Bfi precision

### PR DESCRIPTION
**Pull request by MatzeB**
I noticed that we often have poor precision in the `BlockFrequencyInfo` results because of how `convertFloatingToInteger` in the algorithm chooses factors.

This is a request for comments for choosing more aggressive factors! Look at changes in `test/Analysis/BlockFrequencyInfo/precision.ll` to get an impression for the poor precision and improvements.

Has this been tried before? Is there history I am not aware of?

Right now this change does not work yet as it triggers overflows and some instabilities in various places that I am working through. I thought it may be interesting to get some early feedback though in case this change is miguided...

### CODE REVIEW

**Comment by dexonsmith** [comment]
Comment in the code from the Phab review, which I think was committed mostly as is (https://reviews.llvm.org/D3125, but it wasn't used; the active review was on the commits list):
```
  // Scale the Factor to a size that creates integers.  Ideally, integers would
  // be scaled so that Max == UINT64_MAX so that they can be best
  // differentiated.  However, the register allocator currently deals poorly
  // with large numbers.  Instead, push Min up a little from 1 to give some
  // room to differentiate small, unequal numbers.
  //
  // TODO: fix issues downstream so that ScalingFactor can be Float(1,64)/Max.
```
My memory is that I had equivalent code to this change originally, then backed off to this compromise after hitting performance regressions, but the intention was (originally) to figure out what was wrong we could go back to this.

I don't have access to my email from the time, but I dug up the start of the llvm-commits thread at <https://lists.llvm.org/pipermail/llvm-commits/Week-of-Mon-20140303/207695.html>. Unfortunately it's split by week. I'll reply again if I can find the specific motivation in the thread.

There's been a lot of work on BFI since then so it's possible this will work now.

**Comment by dexonsmith** [comment]
> I don't have access to my email from the time, but I dug up the start of the llvm-commits thread at https://lists.llvm.org/pipermail/llvm-commits/Week-of-Mon-20140303/207695.html. Unfortunately it's split by week. I'll reply again if I can find the specific motivation in the thread.

Actually, it must have come up when I ran tests before posting the patches to the list, since the [initial patch](https://lists.llvm.org/pipermail/llvm-commits/attachments/20140307/778b1bf1/attachment-0005.obj) has the same comment. Almost 10 years ago now, so I can't remember if this was the llvm unit tests or the test-suite.

IMO it's a good idea to revisit this!

**Comment by MatzeB** [comment]
Thanks for confirming @duncan!

The current state is mostly a result of http://reviews.llvm.org/D8718 it seems and I am trying to understand the intentions there or whether that could have been an accident.

I am definitely running into overflow and noise problems with this patch applied and am working on fixing these issues as I see them, hopefully this is manageable.

Another discussion here is whether to normalize "min" value to integer 1 and have higher precisions for the small numbres while saturating big numbers. Or normalize "max" to `UINT64_MAX` which I do here, which may map multiple smaller values to integer 1. I am still collecting experiences here, but my gut feeling would be that we rather want to differentiate the big values than the small ones...

**Comment by david-xl** [comment]
> I noticed that we often have poor precision in the `BlockFrequencyInfo` results because of how `convertFloatingToInteger` in the algorithm chooses factors.
> 
> This is a request for comments for choosing more aggressive factors! Look at changes in `test/Analysis/BlockFrequencyInfo/precision.ll` to get an impression for the poor precision and improvements.
> 
> Has this been tried before? Is there history I am not aware of?
> 
> Right now this change does not work yet as it triggers overflows and some instabilities in various places that I am working through. I thought it may be interesting to get some early feedback though in case this change is miguided...

For precision.ll test case, without the patch, we get 388/600 vs 400/600, and 88/288 vs 100/300. How much do we depend on this level of difference? Profile imprecision is an inherent attribute due to other reasons as well for instance race conditions in instrumentation counter update, sampling period difference etc.

**Comment by MatzeB** [comment]
I don't know if the low precision is a problem for practical code. So far it was mostly annoying me when writing simple testcases that look so obvious that I expected them to just work :) It is also easy to fix those just by choosing a bigger factor there.

Are there reasons why smaller factors may be beneficial?

**Comment by david-xl** [comment]
> I don't know if the low precision is a problem for practical code. So far it was mostly annoying me when writing simple testcases that look so obvious that I expected them to just work :)
> 
> It is also very easy to fix these cases just by choosing bigger factors in `convertFloatingToInteger`...
> 
> Are there any downsides to bigger factors? I am certainly experiencing overflow problems, but those seem worth fixing anyway giving we could also hit them for big min/max spreads in practice...

For PGO, we don't care about two things: 1) ratio of the branch weights of the same branch 2) the 'actual' runtime count of a block or callsite. The later is computed by multiplying the relative frequency (to the entry block) with the entry count (real count). In other words, as long as there is no overflow/underflow, the scale factor used does not really matter.

On the other hand, if overflow happens, the branch direction can be flipped, which is a real problem.

**Comment by MatzeB** [comment]
Thanks for the comment.

I still think choosing more aggressive factors in case of small spreads should not hurt anything and we can easily do so. Even if that just makes people writing test cases happier :)

As for overflows I am still trying to get a feeling for how much we need to go and fix those (as in https://github.com/llvm/llvm-project/pull/66280 ) or whether we have to accept them as a factor of (engineering-)life and rather leave enough room in the values...

**Comment by david-xl** [comment]
> Thanks for the comment.
> 
> I still think choosing more aggressive factors in case of small spreads should not hurt anything and we can easily do so. Even if that just makes people writing test cases happier :)
> 
> As for overflows I am still trying to get a feeling for how much we need to go and fix those (as in #66280 ) or whether we have to accept them as a fact of (engineering-)life and rather leave enough room in the values...

Sure. As long as there are no over/under flow with the new approach, it is a good to have the extra precision kept.

**Comment by MatzeB** [comment]
So we probably want something like the following to make this explicit and have a dial between precision and overflow-bug-avoidance:
```
...
uint64_t Slack = 8;   
uint64_t MaxBits = 64 - Slack;   // We leave some "slack" room to avoid accidental overflows
Scaled64 ScalingFactor = Scaled64(1, MaxBits) / Max;
...
```

I'll try how this goes...

**Comment by qcolombet** [comment]
FWIW, I'm supportive of the investigation here.

Just one comment:
> I am still collecting experiences here, but my gut feeling would be that we rather want to differentiate the big values than the small ones...

I think so, but we probably will lose some valuable information too if we collapse small values.
I expect this will affect hot/cold placement if e.g., two cold blocks have the same weight now and ditto for spill placement in regalloc. I guess it really depends how we "collapse" frequencies (e.g., do we lose 5% accuracy or do we lose 2x for these). In any case, the cases where this matters and overflow occurs are hopefully going to be rare.

Long story short, we'll have regressions :).

**Comment by MatzeB** [comment]
I found that `BlockFrequency` saturation happens a lot in practice. Trying an IPGO enabled service here, I saw hundreds of functions hitting the saturation. This was mostly due to the fact that even nesting two `try/catch` statements pushes the minimum block frequency so far below 1.0 that the current `ScalingFactor = Min.inverse();` even leads to overflow/MAX_UINT saturation on the entry block. To make matters worse I noticed the register allocator often multiplies frequencies with instruction costs and for those cases we loose all of this because the numbers are already saturated. I am measuring a real performance improvement (0.3-0.4%) in this service when I switch to code to the variant above and add some bits (10) as slack so the register allocator and other passes can multiply without triggering immediate saturation.

> In any case, the cases where this matters and overflow occurs are hopefully going to be rare.

If the spread between "min" and "max" frequency is bigger than 64bit then we have to "collapse" on one end (short of replacing the whole system with something like llvm::ScaledNumber or floating point; but that is a discussion for another day I suppose). In the current system I rather collapse the lower-end frequencies to 1 than collapsing the higher end frequencies to MAX_UINT64.


Unfortunately making this change produces a lot of test-case churn. I am still working on this, but published a PR https://github.com/llvm/llvm-project/pull/67197 to increase stability in MachineBlockPlacement against small BFI noise.

**Comment by david-xl** [comment]
@mtrofin the issue (saturation) issue may affect ML based register allocation (training quality).

**Comment by MatzeB** [comment]
no idea how stacked reviews work at the moment. So I included  #67197 in the commits list here for now, note that this is meant to be reviewed and landed before this.

**Comment by WenleiHe** [comment]
> I found that BlockFrequency saturation happens a lot in practice. Trying an IPGO enabled service here, I saw hundreds of functions hitting the saturation. 

Surprising but good find! EH landing pad is almost guaranteed to increase the spread a lot because neither IRPGO nor CSSPGO bother to instrument it. 

> To make matters worse I noticed the register allocator often multiplies frequencies with instruction costs and for those cases we loose all of these costs because the numbers are already saturated.

I'm wondering if we can do more to catch some of such cases with operator* overloads for BlockFrequency and assertions in it.

> In the current system I rather collapse the lower-end frequencies to 1 than collapsing the higher end frequencies to MAX_UINT64.

+1

**Review by WenleiHe — COMMENTED** [review]
The change in convertFloatingToInteger looks good to me. 

This PR has everything including the block placement tolerance change, so I'm not sure how to accept just the BFI precision changes...

File: llvm/test/CodeGen/X86/fsafdo_test3.ll [hunk 1]
```diff
@@ -43,51 +43,51 @@
...
+; BFI:  - BB9[if.end9.1]: float = 59.967, int = {{.*}}, count = 255941
+; BFI:  - BB10[if.then.2]: float = 2.5405, int = {{.*}}, count = 10843
+; BFI:  - BB11[if.end.2]: float = 59.967, int = {{.*}}, count = 255941
+; BFI:  - BB12[if.then7.2]: float = 2.5405, int = {{.*}}, count = 10843
+; BFI:  - BB13[if.end9.2]: float = 59.967, int = {{.*}}, count = 255941
+; BFI:  - BB14[if.then.3]: float = 2.5405, int = {{.*}}, count = 10843
+; BFI:  - BB15[if.end.3]: float = 59.967, int = {{.*}}, count = 255941
+; BFI:  - BB16[if.then7.3]: float = 2.5405, int = {{.*}}, count = 10843
+; BFI:  - BB17[if.end9.3]: float = 59.967, int = {{.*}}, count = 255941
+; BFI:  - BB18[for.end12]: float = 1.0, int = {{.*}}, count = 4268
```
Comment (RKSimon) [inline llvm/test/CodeGen/X86/fsafdo_test3.ll:64]: Why did you hide the int costs?

Comment (MatzeB) [reply llvm/test/CodeGen/X86/fsafdo_test3.ll:64]: They feel a bit more of an implementation detail of the BFI pass. If we happen to further tune their calculation in the future they will change again. While the "float" and "count" value depends on the source alone and should be more stable.
That said no strong opinion, happy to put the costs back if you think that is better.

**Comment by david-xl** [comment]
> I found that `BlockFrequency` saturation happens a lot in practice. Trying an IPGO enabled service here, I saw hundreds of functions hitting the saturation. This was mostly due to the fact that even nesting two `try/catch` statements pushes the minimum block frequency so far below 1.0 that the current `ScalingFactor = Min.inverse();` even leads to overflow/MAX_UINT saturation on the entry block. To make matters worse I noticed the register allocator often multiplies frequencies with instruction costs and for those cases we loose all of these costs because the numbers are already saturated. I am measuring a real performance improvement (0.3-0.4%) in this service when I switch to the code from my comment above (with 10 slack bits) so the register allocator and other passes can multiply without triggering immediate saturation.
> 
> > In any case, the cases where this matters and overflow occurs are hopefully going to be rare.
> 
> If the spread between "min" and "max" frequency is bigger than 64bit then we have to "collapse" on one end (short of replacing the whole system with something like llvm::ScaledNumber or floating point; but that is a discussion for another day I suppose). In the current system I rather collapse the lower-end frequencies to 1 than collapsing the higher end frequencies to MAX_UINT64.
> 
> Unfortunately making this change produces a lot of test-case churn. I am still working on this, but published a PR #67197 to increase stability in MachineBlockPlacement against small BFI noise.

Do you notice a lot of saturation without EH?

File: llvm/test/CodeGen/X86/fsafdo_test3.ll (hunk 1 above)
Comment (RKSimon) [reply llvm/test/CodeGen/X86/fsafdo_test3.ll:64]: OK with the change - thanks

**Comment by MatzeB** [comment]
rebased

**Comment by MatzeB** [comment]
BTW @xur-llvm  could you take a look at `test/Transforms/PGOProfile/bfi_verification.ll`: It seems you added a flag to quantify/print the inaccuracies present in BFI data and well... with my change here things aren't inaccurate anymore and the test no longer prints anything. I am not sure what to do with this except disable the test for now...

**Comment by david-xl** [comment]
> BTW @xur-llvm could you take a look at `test/Transforms/PGOProfile/bfi_verification.ll`: It seems you added a flag to quantify/print the inaccuracies present in BFI data and well with my change here things aren't inaccurate anymore and the test no longer prints anything. I am not sure what to do with this except disable the test for now...

This test catches a diff close 4%. Is it intended to be considered ok with this patch?

To save the test, it is possible to modify the profile data fed to the test case with smaller PGO block counts.

**Comment by MatzeB** [comment]
* Rebased on github/main
* Removed dependency on https://github.com/llvm/llvm-project/pull/67197 (updated more tests instead)
* Partly revived `test/Transforms/PGOProfile/bfi_verification.ll`: I can provoke some count mismatches by cranking up the numbers; though so far I cannot find a way to still provoke the situation where a block flips between hot/cold (as BFI chooses to loose precision on the low-end with this patch but hot/cold decisions happen at the high-end...)

**Review by WenleiHe — APPROVED** [review]
LGTM, assuming tests all pass.

**Comment by nikic** [comment]
It looks like this has made clang 0.7% slower (in the sense that the clang stage2 binary is optimized worse -- there is no negative impact on stage1). The binary is also 0.7% larger. It looks like this change has significant impact on inlining heuristics.

Not really sure what to make of that, I assume you determined that this ends up being faster on average, so maybe clang is just unlucky.

**Comment by MatzeB** [comment]
The internal services I tried here tended to slightly improve or stay neutral.

We can revert, but could you please share some details on how your clang built / tested so I have a chance to address things and bring these changes back?

**Comment by nikic** [comment]
This is a standard release build. The exact cmake config is https://gist.github.com/nikic/1bf97b5b6aade298b87f41cd13bf37d5 where the stage1 build has the same configuration but is built with gcc. I'm testing CTMark, but the impact is quite consistent across the board, so anything will probably do.

**Comment by MatzeB** [comment]
Oh so this isn't even a PGO enabled build, interesting...

**Comment by WenleiHe** [comment]
> This is concerning. Can this be reverted for now and we can help with some internal performance testing. @xur-llvm

Internally, with this change, on a few large workloads, we saw ~0.2-0.5% performance improvements, all with 2-3% .text size reduction at the same time. Our testing covered both sampling pgo and instrumentation pgo.

The change makes profile more accurate, which is a good improvement by itself. But PGO performance isn't just about profile accuracy, and it depends on training quality and representativeness too. Hypothetically, if some of the training isn't that representative relative to perf testing scenario, by making profile more accurate, we could also make profile deviate more from what's needed for best perf on testing scenario. 

That said, I think it'd be fair to revert if this is proven to be a general regression. For that, can we wait for your results before reverting?

**Comment by david-xl** [comment]
Yes -- the revert can wait until more data is available. I agree that it should help performance in theory.

**Comment by nikic** [comment]
> Oh so this isn't even a PGO enabled build, interesting...

Ah yes, I should have mentioned that. This is indeed without PGO. So I expect this is related to the impact of the estimated BFI on the inlining cost model (which is the only significant place using BFI in non-PGO builds). (I've seen issues with bad inlining decisions due to BF underflow in the past, not sure whether this change makes that better or worse.)

**Comment by MatzeB** [comment]
I think I traced the clang regression down to `InlineCostCallAnalyzer::isColdCallSite` returning `true` in more instances. It previously did not do that because of what feels more like an accidental loss of precision:

The example I analyzed starts out as a single-block function with a call resulting in this BFI:
```
 - entry: float = 1.0, int = 8
 ```
 
 It seems that when we inline calls we just scale their BFI values to the callsite. And given that the entry block only had a factor of `int = 8` we have basically no precision left on the low-end. So after inlining a bit more the BFI in my example looks like this:
 ```
  - entry: float = 1.0, int = 8
 - if.then.i: float = 0.0, int = 0
 - if.else.i: float = 0.0, int = 7
 - if.end15.sink.split.i: float = 0.0, int = 0
 - if.end15.i: float = 0.0, int = 8
 - if.then19.i: float = 0.0, int = 5
 - _ZN4llvm12DenseMapBaseINS_8DenseMapIPNS_10StructTypeEPNS_12StructLayoutENS_12DenseMapInfoIS3_vEENS_6detail12DenseMapPairIS3_S5_EEEES3_S5_S7_SA_E20InsertIntoBucketImplIS3_EEPSA_RKS3_RKT_SE_.exit: float = 0.0, int = 8
 ```
 Note that some blocks frequencies got scaled down to `0`. But more importantly no matter what we do it is impossible to have any code marked as "cold" in this situation. Because the cold code threshold is computed as 2% of the entry block frequency. With an entry block value of `int = 8` the 2% cold-threshold is integer `0` and it is effectively impossible for any block to be below that threshold.
 
 
 In comparison with the new conversion scheme we end up with:
 ```
  - entry: float = 1.0, int = 18014398509481984
 - if.then.i: float = 0.0, int = 9002696048640
 - if.else.i: float = 0.0, int = 18005395813433344
 - if.end15.sink.split.i: float = 0.0, int = 18000892999733
 - if.end15.i: float = 0.0, int = 18014398509481984
 - if.then19.i: float = 0.0, int = 11258999068426240
 - _ZN4llvm12DenseMapBaseINS_8DenseMapIPNS_10StructTypeEPNS_12StructLayoutENS_12DenseMapInfoIS3_vEENS_6detail12DenseMapPairIS3_S5_EEEES3_S5_S7_SA_E20InsertIntoBucketImplIS3_EEPSA_RKS3_RKT_SE_.exit: float = 0.0, int = 18014398509481984
 ```
for the same situation and blocks like `if.end15.sink.split.i` end up being classified as cold code now.

**Comment by MatzeB** [comment]
So not sure if there is an easy fix for the regression.

We should probably have some logic that scales all BFI values when inlining to better accomodate the new range of frequencies so we loose less precision by accident. Though that won't help with the regression at hand as that would need for `isColdCallSite` to accidentally not work as before to get more inlining...

**Comment by MatzeB** [comment]
Also note that this 2% cold-code threshold is only applied in situation where no PGO data is available. So maybe we should just ignore this, given that without PGO data it just is often impossible for the compiler to make good inlining decisions...

**Comment by dexonsmith** [comment]
Seems awkward to pessimize as "cold" when there's no real data (no PGO, no other marking). What happens if you change the cold call threshold to 0% when there's no PGO data? (I.e., never assume a call is cold without actual evidence)

**Comment by dexonsmith** [comment]
Not entirely accidental. When BPI/BFI first landed it was heavily profiled to be sure it didn't pessimize non-PGO code. I don't see why we'd suddenly be okay with pessimizing it.

Under 2% isn't hard to hit for hot path code. Lots of functions will have strings of early exit conditions.

**Comment by david-xl** [comment]
Agree that deciding coldness based on local entry frequency (2%) is a bad idea though.

**Comment by MatzeB** [comment]
Seems this got introduced in https://reviews.llvm.org/D34312 with the rough idea that we shouldn't inline into parts of the code that `_builtin_expect(...)` deems unlikely. Which makes sense when you say it express it like this, but I guess numeric thresholds can go wrong...

**Comment by dexonsmith** [comment]
> Seems this got introduced in https://reviews.llvm.org/D34312 with the rough idea that we shouldn't inline into parts of the code that `_builtin_expect(...)` deems unlikely. Which makes sense when you express it like this, but I guess numeric thresholds can go wrong...

Heh, yeah, the premise seems correct, but a percentage-based numeric threshold doesn't seem right. You kind of want a flag for the block. Or a special value, like "freq=0", which indicates "annotation says this is cold".

**Comment by minglotus-6** [comment]
> Seems this got introduced in https://reviews.llvm.org/D34312 with the rough idea that we shouldn't inline into parts of the code that _builtin_expect(...) deems unlikely

@xur-llvm @david-xl This just reminds me of the observed regressions when we did an experiment to override branch probability from `absl_predict_{true,false}` with (manually collected IIRC) sampled PGO profiles.

**Comment by david-xl** [comment]
Contracting my past self (one of the reviewers of the patch),  I think coldness check should be based on a global threshold, which exists when synthetic entry count propagation is enabled (without PGO).

**Comment by vitalybuka** [comment]
FYI @ldionne 

This patch causing test_vector2.pass.cpp trigger memory leak https://lab.llvm.org/buildbot/#/builders/168/builds/16422
Not sure if this was leaking but lsan didn't see, it's possible. Or some regression in lsan.

**Comment by MatzeB** [comment]
Not sure how this change could trigger leaks. Changing BFI here should mostly effect inlining decisions, basic block placement, register allocation. But it shouldn't change program behavior or memory operations...

Will try to reproduce this test on my machine, to see if there's anything obvious for the change here (though at a first glance the test code seems to provoke an exception abort and would naturally leak because of it?)

**Comment by MatzeB** [comment]
Some very ad-hoc benchmarking. Of clang compilation speed (measured in instructions as reported by valgrind/callgrind which I think somewhat matches the setup of nikic) compiling `sqlite3` of CTMark:

Old-BFI (this PR reverted), New-BFI (this PR applied), no-cold (cold-callsite-rel-freq set to 0 to disable `InlineCostCallAnalyzer::isColdCallSite` behavior for non-PGO builds:

```
Old-BFI:          size:  59,802,772  (baseline)  insn: 2,812,833,144  (baseline)
New-BFI:          size:  60,247,076  +0.74%      insn: 2,818,639,641  +0.21%
Old-BFI, no-cold: size:  60,773,988  +1.62%      insn: 2,806,521,932  -0.22%
New-BFI, no-cold: size:  60,741,700  +1.57%      insn: 2,803,489,298  -0.33%
```

I could benchmark more with llvm-test-suite and put up a PR to disable cold-callsite-rel-freq by default if people support this direction...

**Comment by MatzeB** [comment]
And digging even deeper:
- FWIW I noticed that I only used `clang -c` as benchmark previously, should have used `clang -c -O3` resulting in this:
```
Old-BFI:          insn: 37,821,687,947  (baseline)
New-BFI:          insn: 38,133,312,923  +0.82%
Old-BFI, no-cold: insn: 37,423,365,184  -1.05%
New-BFI, no-cold: insn: 37,438,736,713  -1.01%
```
- The problematic part of the code that is inlined/not-inlined is actually marked as `LLVM_UNLIKELY` here: https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ADT/DenseMap.h#L607  and intuitively one would think that it is probably best to not inline `grow`  (as will happen with the new BFI version)
- In practice not-inlining `grow` mostly ends up being worse because of knock-on effects as far as I can tell. These are the inlining decisions I noticed for the most prominent situation:
    - `grow` inlined (Old-BFI):
         - Instruction::getMetadataImpl
             -> Value::getMetadata           not inlined
             -> DenseMap::operator[]         inlined
             -> DenseMap::FindAndConstruct   inlined
             -> DenseMap::InsertIntoBucket   not inlined, size
                   likely too big with `grow` inlined here
             -> DenseMap::grow               inlined
    - `grow` inlined (New-BFI):
         - Instruction::getMadataImpl
             -> Value::getMetadata          inlined
             -> DenseMap::operator[]        inlined
             -> DenseMap::FindAndConstruct  not inlined, size
             -> DenseMap::InsertIntoBucket  inlined
             -> DenseMap::grow              not inlined
             
Though if you look into this then I would state that the code got better for the wrong reasons! Not inlining `grow` is a sensible decision in this context and the `LLVM_UNLIKELY` annotation makes sense (I actually added some counters and see the unlikely branch taken somewhere between 1% and 10% of the cases depending on inputs, so seems sensible).

Unfortunately the particular code here `getMetadataImpl` never inserts new things into the map, but unfortunately `operator[]` gives you that behavior by default so nobody noticed. So not inlining `InsertIntoBucket` happened to be a great decision previously that the compiler did by accident without having good data to support this. Now with better but still insufficient data (as this is PGO) we happen to make the decisions that ends up being worse.

Long story short: This ended up another of the many stories where the compiler cannot make the right inlining decisions without having actual runtime data. It tends to make a better decision based on the data that ends up being wrong anyway here. I'm gonna leave things as is and rather put up some improvements to LLVM code instead!

**Comment by dexonsmith** [comment]
Interesting. Probably `Value::getMetadata()` could/should call `DenseMap::find()` instead of `operator[]()` and assert that it's found before dereferencing, because `Value::hasMetadata()` (which, IIRC, consults a bit stored in `Value`) has already promised something will be there. Probably you'll find me on the git-blame for this...

Your plan SGTM!

**Comment by david-xl** [comment]
> And digging even deeper:
> 
> * FWIW I noticed that I only used `clang -c` as benchmark previously, should have used `clang -c -O3` resulting in this:
> 
> ```
> Old-BFI:          insn: 37,821,687,947  (baseline)
> New-BFI:          insn: 38,133,312,923  +0.82%
> Old-BFI, no-cold: insn: 37,423,365,184  -1.05%
> New-BFI, no-cold: insn: 37,438,736,713  -1.01%
> ```
> 
> * The problematic part of the code that is inlined/not-inlined is actually marked as `LLVM_UNLIKELY` here: https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ADT/DenseMap.h#L607  and intuitively one would think that it is probably best to not inline `grow`  (as will happen with the new BFI version)
> * In practice not-inlining `grow` mostly ends up being worse because of knock-on effects as far as I can tell. These are the inlining decisions I noticed for the most prominent situation:
> 
> ```
>     - `grow` inlined (Old-BFI):
>          - Instruction::getMetadataImpl
>              -> Value::getMetadata           not inlined
>              -> DenseMap::operator[]         inlined
>              -> DenseMap::FindAndConstruct   inlined
>              -> DenseMap::InsertIntoBucket   not inlined, size
>                    likely too big with `grow` inlined here
>              -> DenseMap::grow               inlined
>     - `grow` inlined (New-BFI):
>          - Instruction::getMadataImpl
>              -> Value::getMetadata          inlined
>              -> DenseMap::operator[]        inlined
>              -> DenseMap::FindAndConstruct  not inlined, size
>              -> DenseMap::InsertIntoBucket  inlined
>              -> DenseMap::grow              not inlined
> ```
> 
> Though if you look into this then I would state that the code got better for the wrong reasons! Not inlining `grow` is a sensible decision in this context and the `LLVM_UNLIKELY` annotation makes sense (I actually added some counters and see the unlikely branch taken somewhere between 1% and 10% of the cases depending on inputs, so seems sensible).
> 
> Unfortunately the particular code here `getMetadataImpl` never inserts new things into the map, but unfortunately `operator[]` gives you that behavior by default so nobody noticed. So not inlining `InsertIntoBucket` happened to be a great decision previously that the compiler did by accident without having good data to support this. Now with better but still insufficient data (as this is PGO) we happen to end up inlining `InsertIntoBucket` wasting code size leading to worse inlining decisions further up the stack...
> 
> Long story short: This ended up another of the many stories where the compiler cannot make the right inlining decisions without having actual runtime data. It tends to make a better decision based on the data that ends up being wrong anyway here. I'm gonna leave things as is and rather put up some improvements to LLVM code instead!

Good analysis.

Your plan sounds good but lowering the cold threshold is probably a good thing to do with the new BFI. The FindAndConstruct method in DenseMap can probably be annotated with buitin_expect to indicate the insertion is a unlikely path.

**Comment by WenleiHe** [comment]
Thanks for investigating, Matthias. So it looks like the regression is indeed just clang non-PGO build being unlucky after the BFI improvement.. Sounds good to leave it as is.

**Comment by MatzeB** [comment]
https://github.com/llvm/llvm-project/pull/70700  improves LLVM code so I am no longer seeing a regression before/after this change.

**Comment by david-xl** [comment]
@MatzeB ,Our internal release testing have seen lots of very large regressions for tests without PGO or with XFDO only. While I agree this patch is the right way to go and the good performance (without PGO) with oldBFI is somewhat by chance, it is still better not to regress them. Can you help with a follow up patch to tune down relative coldness threshold?

**Comment by MatzeB** [comment]
Do you just see regression or also some wins? I don't know whether this is all stems from the `isColdCallSite` behavior, and not just the usual some things get better some things get worse situation you can easily have when inlining changes?

**Comment by MatzeB** [comment]
Or put another way: Do you see regressions disappear with `-mllvm -cold-callsite-rel-freq=0` ?

**Comment by david-xl** [comment]
I asked for those data points and share it when ready.
